src/encode/n_axioms.ml

```diff
@@ -0,0 +1,5029 @@
+        ; apps T.IntSet [] %% []
+        ] %% []
+      ; apps T.Mem
+        [ Ix 1 %% []
+        ; apps T.IntSet [] %% []
+        ] %% []
+      ] %% []
+    ; apps T.Mem
+      [ apps T.IntPlus
+        [ Ix 2 %% []
+        ; Ix 1 %% []
+        ] %% []
+      ; apps T.IntSet [] %% []
+      ] %% []
+    ] %% []
+  ) %% []
+
+let intuminus_typing () =
+  quant Forall
+  [ "x" ] [ t_idv ]
+  ~pats:[ [
+    apps T.IntUminus
+    [ Ix 1 %% []
+    ] %% []
+  ] ]
+  ( appb B.Implies
+    [ apps T.Mem
+      [ Ix 1 %% []
+      ; apps T.IntSet [] %% []
+      ] %% []
+    ; apps T.Mem
+      [ apps T.IntUminus
+        [ Ix 1 %% []
+        ] %% []
+      ; apps T.IntSet [] %% []
+      ] %% []
+    ] %% []
+  ) %% []
+
+let intminus_typing () =
+  quant Forall
+  [ "x" ; "y" ] [ t_idv ; t_idv ]
+  ~pats:[ [
+    apps T.IntMinus
+    [ Ix 2 %% []
+    ; Ix 1 %% []
+    ] %% []
+  ] ]
+  ( appb B.Implies
+    [ appb B.Conj
+      [ apps T.Mem
+        [ Ix 2 %% []
+        ; apps T.IntSet [] %% []
+        ] %% []
+      ; apps T.Mem
+        [ Ix 1 %% []
+        ; apps T.IntSet [] %% []
+        ] %% []
+      ] %% []
+    ; apps T.Mem
+      [ apps T.IntMinus
+        [ Ix 2 %% []
+        ; Ix 1 %% []
+        ] %% []
+      ; apps T.IntSet [] %% []
+      ] %% []
+    ] %% []
+  ) %% []
+
+let inttimes_typing () =
+  quant Forall
+  [ "x" ; "y" ] [ t_idv ; t_idv ]
+  ~pats:[ [
+    apps T.IntTimes
+    [ Ix 2 %% []
+    ; Ix 1 %% []
+    ] %% []
+  ] ]
+  ( appb B.Implies
+    [ appb B.Conj
+      [ apps T.Mem
+        [ Ix 2 %% []
+        ; apps T.IntSet [] %% []
+        ] %% []
+      ; apps T.Mem
+        [ Ix 1 %% []
+        ; apps T.IntSet [] %% []
+        ] %% []
+      ] %% []
+    ; apps T.Mem
+      [ apps T.IntTimes
+        [ Ix 2 %% []
+        ; Ix 1 %% []
+        ] %% []
+      ; apps T.IntSet [] %% []
+      ] %% []
+    ] %% []
+  ) %% []
+
+let intexp_typing () =
+  quant Forall
+  [ "x" ; "y" ] [ t_idv ; t_idv ]
+  ~pats:[ [
+    apps T.IntExp
+    [ Ix 2 %% []
+    ; Ix 1 %% []
+    ] %% []
+  ] ]
+  ( appb B.Implies
+    [ List (And,
+      [ apps T.Mem
+        [ Ix 2 %% []
+        ; apps T.IntSet [] %% []
+        ] %% []
+      ; appb ~tys:[ t_idv ] B.Neq
+        [ Ix 2 %% []
+        ; apps (T.IntLit 0) [] %% []
+        ] %% []
+      ; apps T.Mem
+        [ Ix 1 %% []
+        ; apps T.IntSet [] %% []
+        ] %% []
+      ]) %% []
+    ; apps T.Mem
+      [ apps T.IntExp
+        [ Ix 2 %% []
+        ; Ix 1 %% []
+        ] %% []
+      ; apps T.IntSet [] %% []
+      ] %% []
+    ] %% []
+  ) %% []
+
+let intquotient_typing () =
+  quant Forall
+  [ "x" ; "y" ] [ t_idv ; t_idv ]
+  ~pats:[ [
+    apps T.IntQuotient
+    [ Ix 2 %% []
+    ; Ix 1 %% []
+    ] %% []
+  ] ]
+  ( appb B.Implies
+    [ List (And,
+      [ apps T.Mem
+        [ Ix 2 %% []
+        ; apps T.IntSet [] %% []
+        ] %% []
+      ; apps T.Mem
+        [ Ix 1 %% []
+        ; apps T.NatSet [] %% []
+        ] %% []
```

Comment:
According to Specifying Systems: "We define ÷ and % so that
a = b ∗ (a ÷ b) + (a % b)
for all integers a and b with b > 0"

It's not entirely clear to me if floor division and remainder are supposed to be defined in TLA+ for negative b. The quote above suggests it's not, so I went with it to be careful.

Reply:
Integer division is defined in module Naturals (cf. p. 348 of Specifying Systems) as
```a \div b == CHOOSE n \in Int : \E r \in 0 .. (b-1) : a = b*n+r```
This implies that integer division is essentially undefined for `b <= 0` (it boils down to `CHOOSE n \in Int : FALSE`). There are basically two ways for defining it when `b` is negative, I don't know which definition SMT chooses, but TLA's `\div` should not be mapped to SMT's operator when `b` is negative.

Since remainder is defined from `\div`, the same applies there.